sql/src/main/java/org/apache/druid/sql/calcite/planner/querygen/DruidQueryGenerator.java

```diff
@@ -0,0 +1,296 @@
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.druid.sql.calcite.planner.querygen;
+
+import com.google.common.collect.ImmutableList;
+import org.apache.calcite.rel.RelNode;
+import org.apache.calcite.rel.core.Aggregate;
+import org.apache.calcite.rel.core.Filter;
+import org.apache.calcite.rel.core.Project;
+import org.apache.calcite.rel.core.Sort;
+import org.apache.calcite.rel.core.Window;
+import org.apache.calcite.rex.RexBuilder;
+import org.apache.druid.error.DruidException;
+import org.apache.druid.query.QueryDataSource;
+import org.apache.druid.sql.calcite.planner.PlannerContext;
+import org.apache.druid.sql.calcite.planner.querygen.DruidQueryGenerator.PDQVertexFactory.PDQVertex;
+import org.apache.druid.sql.calcite.planner.querygen.InputDescProducer.InputDesc;
+import org.apache.druid.sql.calcite.rel.DruidQuery;
+import org.apache.druid.sql.calcite.rel.PartialDruidQuery;
+import org.apache.druid.sql.calcite.rel.PartialDruidQuery.Stage;
+
+import javax.annotation.Nullable;
+
+import java.util.ArrayList;
+import java.util.List;
+
+/**
+ * Converts a DAG of {@link org.apache.druid.sql.calcite.rel.logical.DruidLogicalNode} convention to a native
+ * {@link DruidQuery} for execution.
+ */
+public class DruidQueryGenerator
+{
+  private final RelNode relRoot;
+  private final PDQVertexFactory vertexFactory;
+
+  public DruidQueryGenerator(PlannerContext plannerContext, RelNode relRoot, RexBuilder rexBuilder)
+  {
+    this.relRoot = relRoot;
+    this.vertexFactory = new PDQVertexFactory(plannerContext, rexBuilder);
+  }
+
+  public DruidQuery buildQuery()
+  {
+    Vertex vertex = buildVertexFor(relRoot, true);
+    return vertex.buildQuery(true);
+  }
+
+  private Vertex buildVertexFor(RelNode node, boolean isRoot)
+  {
+    List<Vertex> newInputs = new ArrayList<>();
+    for (RelNode input : node.getInputs()) {
+      newInputs.add(buildVertexFor(input, false));
+    }
+    Vertex vertex = processNodeWithInputs(node, newInputs, isRoot);
+    return vertex;
+  }
+
+  private Vertex processNodeWithInputs(RelNode node, List<Vertex> newInputs, boolean isRoot)
+  {
+    if (node instanceof InputDescProducer) {
+      return vertexFactory.createVertex(PartialDruidQuery.create(node), newInputs);
+    }
+    if (newInputs.size() == 1) {
+      Vertex inputVertex = newInputs.get(0);
+      Vertex newVertex = inputVertex.mergeNode(node, isRoot);
```

Comment:
instead of `mergeNode`, does this seem more like a `addParent`/ `buildParent`? 
Also, the method name can contain a `maybe` too since it can return `null` as well.

Reply:
I've used `Optional` instead `maybe`  - this made me remove some apidoc as well ; as now the returntype tells the story (instead of me in an apidoc)

Since we have a `Vertex` here the use of the keyword `parent` might be misleading.

I've renamed it to `extendWith` ; new method signature:
```
Optional<Vertex> extendWith(RelNode parentNode, boolean isRoot)
```